cache-manager/cache-manager.php

```diff
@@ -0,0 +1,433 @@
+
+		while ( $running ) {
+			do {
+				$result = curl_multi_exec( $curl_multi, $running );
+			} while ( $result == CURLM_CALL_MULTI_PERFORM );
+
+			if ( $result != CURLM_OK )
+				error_log( 'curl_multi_exec() returned something different than CURLM_OK' );
+
+			curl_multi_select( $curl_multi, 0.2 );
+		}
+
+		while ( $completed = curl_multi_info_read( $curl_multi ) ) {
+			$info = curl_getinfo( $completed['handle'] );
+
+			if ( ! $info['http_code'] && curl_error( $completed['handle'] ) )
+				error_log( 'Error on: ' . $info['url'] . ' error: ' . curl_error( $completed['handle'] ) . "\n" );
+
+			if ( '200' != $info['http_code'] )
+				error_log( 'Request to ' . $info['url'] . ' returned HTTP code ' . $info['http_code'] . "\n" );
+
+			curl_multi_remove_handle( $curl_multi, $completed['handle'] );
+		}
+
+		curl_multi_close( $curl_multi );
+	}
+
+	protected function build_purge_request( $url, $method ) {
+		global $varnish_servers;
+
+		$requests = array();
+
+		if ( empty( $varnish_servers ) )
+			return $requests;
+
+		$parsed = parse_url( $url );
+		if ( empty( $parsed['host'] ) )
+			return $requests;
+
+		foreach ( $varnish_servers as $server  ) {
+			$server = explode( ':', $server[0] );
+
+			$uri = '/';
+			if ( isset( $parsed['path'] ) )
+				$uri = $parsed['path'];
+			if ( isset( $parsed['query'] ) )
+				$uri .= $parsed['query'];
+
+			$requests[] = array(
+				'ip'     => $server[0],
+				'port'   => $server[1],
+				'host'   => $parsed['host'],
+				'uri'    => $uri,
+				'method' => $method
+			);
+		}
+
+		return $requests;
+	}
+
+	public function execute_purges() {
+		$this->ban_urls = array_unique( $this->ban_urls );
+		$this->purge_urls = array_unique( $this->purge_urls );
+
+		if ( empty( $this->ban_urls ) && empty( $this->purge_urls ) )
+			return;
+
+		$requests = array();
+		foreach( (array) $this->ban_urls as $url )
+			$requests = array_merge( $requests, $this->build_purge_request( $url, 'BAN' ) );
+
+		foreach( (array) $this->purge_urls as $url )
+			$requests = array_merge( $requests, $this->build_purge_request( $url, 'PURGE' ) );
+
+		$this->ban_urls = $this->purge_urls = array();
+
+		if ( empty( $requests ) )
+			return;
+
+		return $this->curl_multi( $requests );
+	}
+
+	public function purge_site_cache( $when = null ) {
+		if ( $this->site_cache_purged )
+			return;
+
+		$this->ban_urls[] = untrailingslashit( home_url() ) . '/.*';
+		$this->site_cache_purged = true;
+
+		return;
+	}
+
+	public function queue_post_purge( $post_id ) {
+		if ( $this->site_cache_purged )
+			return false;
+
+		if ( defined( 'WP_IMPORTING' ) ) {
+			$this->purge_site_cache();
+			return false;
+		}
+
+		$post = get_post( $post_id );
+		if ( empty( $post ) ||
+		     'revision' === $post->post_type ||
+		     ! in_array( get_post_status( $post_id ), array( 'publish', 'trash' ), true ) )
+		{
+			return false;
+		}
+
+		$this->purge_urls[] = get_permalink( $post_id );
+		$this->purge_urls[] = trailingslashit( home_url() );
```

Comment:
Better to use `home_url( '/' )` instead, since we know we want a trailing slash. `trailingslashit()` adds a bunch of unnecessary processing.